service/frontend/workflowHandler.go

```diff
@@ -3538,18 +3538,42 @@ func (wh *WorkflowHandler) GetDomainReplicationMessages(
 		return nil, wh.error(errors.New("domain replication queue not enabled for cluster"), scope)
 	}
 
-	// TODO: Set it to last ack level for the cluster.
 	lastMessageID := defaultLastMessageID
 	if request.IsSetLastRetrivedMessageId() {
 		lastMessageID = int(request.GetLastRetrivedMessageId())
 	}
 
+	if lastMessageID == defaultLastMessageID {
+		clusterAckLevels, err := wh.domainReplicationQueue.GetAckLevels()
+		if err == nil {
+			if ackLevel, ok := clusterAckLevels[request.GetClusterName()]; ok {
+				lastMessageID = ackLevel
+			}
+		}
+	}
+
 	replicationTasks, lastMessageID, err := wh.domainReplicationQueue.GetReplicationMessages(
 		lastMessageID, getDomainReplicationMessageBatchSize)
 	if err != nil {
 		return nil, wh.error(err, scope)
 	}
 
+	lastProcessedMessageID := defaultLastMessageID
+	if request.IsSetLastProcessedMessageId() {
+		lastProcessedMessageID = int(request.GetLastProcessedMessageId())
+	}
+
+	if lastProcessedMessageID != defaultLastMessageID {
+		go func() {
```

Comment:
why a separate coroutine here

Reply:
Original intention is to ignore checkpointing failures. Talked to @wxing1292 and we will handle it in the RPC handler go routine.